### PR TITLE
[WIP] PIP 35: Improve topic lookup for topics that have high number of partitions

### PR DESCRIPTION
**Pull request by MarvinCai**
[WIP]  PIP 35: Improve topic lookup for topics that have high number of partitions
Pending UT and change in Proxy classes.

Add batch lookup ability to LookupService.
Accept list of topic names.
Whenever possible, return partial succeed result instead of failing whole batch lookup result.

For all topics pass in in request, first decided which NDBundle they belong to, then do lookup on each bundle.
Response will have map of broker address to list of topics that broker serve. 
Failed lookup will be in separate list.

*(If this PR fixes a github issue, please add `Fixes #<xyz>`.)*

Fixes #<xyz>

*(or if this PR is one task of a github issue, please add `Master Issue: #<xyz>` to link to the master issue.)*

Master Issue: #<xyz>

### Motivation

*Explain here the context, and why you're making that change. What is the problem you're trying to solve.*

### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Comment by MarvinCai** [comment]
This is WIP, still have some changes but most of the flow is added.
Want to get some initial comments before adding UT and making change in Proxy classes.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationService.java [hunk 1]
```diff
@@ -305,6 +307,55 @@ public boolean canLookup(TopicName topicName, String role, AuthenticationDataSou
...
+                    unAuthorizedTopicNames.add(topicNames.get(i));
+                }
+            };
+            finalResult.complete(unAuthorizedTopicNames);
+        });
+
+        return finalResult;
+    }
+
+    public List<TopicName> canBatchLookup(List<TopicName> topicNames, String role,
```
Comment (jiazhai) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationService.java]: How about  in this function simply call `canBatchLookupAsync.get()`

File: pulsar-common/src/main/proto/PulsarApi.proto [hunk 2]
```diff
@@ -184,6 +184,7 @@ enum ProtocolVersion {
 	v13 = 13; // Schema-registry : added avro schema format for json
 	v14 = 14; // Add CommandAuthChallenge and CommandAuthResponse for mutual auth
 	          // Added Key_Shared subscription
+	v15 = 15; // Add batch lookup support
```
Comment (sijie) [inline pulsar-common/src/main/proto/PulsarApi.proto]: depend on when this PR will be merged. if it is merged in 2.4.0, you can just add "Add batch lookup support" to v14.

**Comment by sijie** [comment]
@rdhabalia @merlimat since you are familiar with the lookup logic, can you please take a look at this proposal, to make sure @MarvinCai is in a right redirection to continue?

**Comment by jiazhai** [comment]
@MarvinCai Thanks for the work. overall looks good. looking forward to see the final pr.

**Comment by tisonkun** [comment]
Closed as stale and conflict.

**Pull request closed without merging**